File: app-library-starter-core/src/main/java/com/github/arhor/aws/graphql/federation/starter/core/data/FeaturesWritingConverter.java
package com.github.arhor.aws.graphql.federation.starter.core.data;

import org.jetbrains.annotations.NotNull;
import org.springframework.core.convert.converter.Converter;
import org.springframework.data.convert.WritingConverter;

@WritingConverter
public class FeaturesWritingConverter implements Converter<Features<?, ?>, Integer> {

    @NotNull
    @Override
    public Integer convert(@NotNull final Features<?, ?> source) {
        var result = 0;

        for (final var item : source.getItems()) {
            result |= 1 << item.ordinal();
        }
        return result;
    }
}

File: lib-starter-core/src/main/kotlin/com/github/arhor/aws/graphql/federation/starter/core/data/FeaturesWritingConverter.kt
package com.github.arhor.aws.graphql.federation.starter.core.data

import org.springframework.core.convert.converter.Converter
import org.springframework.data.convert.WritingConverter

@WritingConverter
open class FeaturesWritingConverter : Converter<Features<*, *>, Int> {

    override fun convert(source: Features<*, *>): Int =
        source.items.fold(0) { result, item -> result or (1 shl item.ordinal) }
}
